comma/database/driver.py

```diff
@@ -28,42 +32,73 @@ def __init__(self, dry_run, echo=False):
         # Enable INFO-level logging when program is logging debug
         # It's not ideal, because the messages are INFO level, but only enabled with debug
 
+        # As defined in msodbcsql.h
+        self.SQL_COPT_SS_ACCESS_TOKEN = 1256
+
         if dry_run:
             db_file = "comma.db"
             LOGGER.info("Using local SQLite database at '%s'.", db_file)
             engine = sqlalchemy.create_engine(f"sqlite:///{db_file}", echo=echo)
         else:
             LOGGER.info("Connecting to remote database...")
-            engine = sqlalchemy.create_engine(self._get_mssql_conn_str(), echo=echo)
+            engine = self.create_engine("ODBC Driver 17 for SQL Server")
 
         Base.metadata.bind = engine
         Base.metadata.create_all(engine)
         self.session_factory = sqlalchemy.orm.sessionmaker(bind=engine)
 
-    @staticmethod
-    def _get_mssql_conn_str() -> sqlalchemy.engine.Engine:
-        """
-        Create a connection string for MS SQL server and create engine instance
-        """
+    def get_token(self) -> bytes:
+        try:
+            credential = DefaultAzureCredential()
+            cre_token = credential.get_token("https://database.windows.net/").token
+            token = cre_token.encode("utf-16-le")
+            token_struct = struct.pack(f"=I{len(token)}s", len(token), token)
+            return bytes(token_struct)
+        except Exception as e:
+            raise RuntimeError("Failed to obtain Azure AD token") from e
+
+    def create_engine(self, driver: str) -> Any:
+        if os.environ.get('COMMA_DB_USERNAME') and os.environ.get('COMMA_DB_PW'):
+            return self.create_engine_with_pass(driver)
+        else:
+            return self.create_engine_with_token(driver)
```

Comment:
what are the use cases of using signing with token?
Also, if we are hardcoding TOKEN is that secured? Should we use env variables?

Reply:
New security standards no longer allow password access to SQL servers. We are not hard coding TOKEN. We get token from the azure.identity function DefaultAzureCredential(). This function looks to see if AzCLI is signed in and also checks a few other places where azure identity credentials are stored.

Reply:
https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet